internal/oauth/oauth.go

```diff
@@ -0,0 +1,190 @@
+package oauth
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/foundriesio/fioctl/internal/config"
+	"github.com/sirupsen/logrus"
+)
+
+const uri string = "https://app.foundries.io/oauth"
+
+// OAuth interface.
+type OAuth interface {
+	Get() int
+	HasRefreshToken() bool
+	IsExpired() (bool, error)
+	Refresh() int
+	post(string, url.Values) (*[]byte, error)
+}
+
+// The ClientCredentials struct implements the OAuth interface for the
+// client_credentials/refresh_token OAuth grant types.
+type ClientCredentials struct {
+	Config *config.Config
+	URL    string
+}
+
+type OAuthResponse struct {
+	TokenType    string  `json:"token_type"`
+	AccessToken  string  `json:"access_token"`
+	RefreshToken string  `json:"refresh_token"`
+	ExpiresIn    float64 `json:"expires_in"`
+	Scope        string  `json:"scope"`
+}
+
+func buildUrl(uri string, p string) (string, error) {
+	u, err := url.Parse(uri)
+	if err != nil {
+		return "", err
+	}
+
+	u.Path = path.Join(u.Path, p)
+	if !strings.HasSuffix(u.Path, "/") {
+		u.Path += "/"
+	}
+
+	return u.String(), nil
+}
+
+func updateConfig(c *config.Config, r *OAuthResponse) {
+	c.ClientCredentials.AccessToken = r.AccessToken
+	c.ClientCredentials.RefreshToken = r.RefreshToken
+	c.ClientCredentials.TokenType = r.TokenType
+	c.ClientCredentials.ExpiresIn = r.ExpiresIn
+	c.ClientCredentials.Created = time.Now().UTC().Format(time.RFC3339)
+}
+
+// Perform a POST request.
+func (c *ClientCredentials) post(uri string, data url.Values) (*[]byte, error) {
+	res, err := http.PostForm(uri, data)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if res.StatusCode != 200 {
+		return nil, fmt.Errorf("Unable to retrieve credentials: HTTP_%d\n=%s", res.StatusCode, body)
+	}
+
+	return &body, nil
+}
+
+// IsExpired checks if the OAuth token is expired.
+func (c *ClientCredentials) IsExpired() (bool, error) {
+	if c.Config.ClientCredentials.ExpiresIn != 0 && c.Config.ClientCredentials.Created != "" {
+		t, err := time.Parse(time.RFC3339, c.Config.ClientCredentials.Created)
+		if err != nil {
+			return false, fmt.Errorf("Error parsing oauth token creation date")
+		}
+
+		if time.Since(t).Seconds() > c.Config.ClientCredentials.ExpiresIn {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// HasRefreshToken checks if a refresh token is available.
+func (c *ClientCredentials) HasRefreshToken() bool {
+	if c.Config.ClientCredentials.RefreshToken != "" {
+		return true
+	}
+	return false
```

Comment:
linter caught this: you can just do `return c.Config.ClientCredentials.RefreshToken != ""` pretty decent linting logic.

Reply:
Good to know. Have to check VS Code go linter better, it didn't report that.